### PR TITLE
Package check

### PR DESCRIPTION
**Pull request by MarquiseRosier**
Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues
#167 

Thanks for contributing!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -85,8 +87,15 @@
     "@adobe/helix-status": "7.1.2",
     "@adobe/openwhisk-action-logger": "2.2.0",
     "@adobe/openwhisk-action-utils": "4.2.2",
+    "cz-conventional-changelog": "^3.1.0",
     "fs-extra": "9.0.0",
     "glob-to-regexp": "0.4.1",
+    "openwhisk": "^3.21.1",
```
Comment (tripodsan) [inline package.json:93]: why is this a dev dependency?

File: src/check-pkgs.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * @param namespace {string} openwhisk namespace under which actions exist
+ */
+function check_pkgs(auth, host, namespace, config) {
+  return new Promise((resolve, reject) => {
+    const ow = openwhisk({
+      api_key: auth,
+      apihost: host,
+      namespace,
+    });
+    ow.packages.list().then((data) => {
```
Comment (tripodsan) [inline src/check-pkgs.js:26]: why use promises and not async/await ?

File: src/check-pkgs.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      const package_list = data.reduce((prev, curr) => {
+        prev[curr.name] = true;
+        return prev;
+      }, {});
+
+      config.strains.forEach((strain) => {
+        if (strain.package && !(strain.package in package_list)) {
+          return reject(Error(`config package for the strain: ${strain.name} not deployed`));
```
Comment (tripodsan) [inline src/check-pkgs.js:34]:
```suggestion
          return reject(Error(`action package ${strain.package} for the strain: ${strain.name} not deployed`));
```

File: src/publish.js [hunk 4]
```diff
@@ -29,7 +29,7 @@ const redirects = require('./fastly/redirects');
  * @param {object} iconfig the IndexConfig from helix-shared
  * @param {string} algoliaappid Algolia App ID (not secret)
  */
-async function publish(configuration, service, token, version, vclOverrides = {}, dispatchVersion = 'v3', log = console, iconfig, algoliaappid) {
+async function publish(configuration, service, token, version, vclOverrides = {}, dispatchVersion = 'v3', log = console, iconfig, algoliaappid, wskAuth, wskHost, wskNamespace) {
```
Comment (tripodsan) [inline src/publish.js:32]: this gets longer and longer....maybe switch to a different style. eg:
- pass a `options` object
- use a builder pattern

(but this can be addresses in a different PR)

File: test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har [hunk 5]
```diff
@@ -4930,6 +4930,105 @@
...
+            {
+              "name": "host",
+              "value": "fake_host"
+            }
+          ],
+          "headersSize": 347,
+          "httpVersion": "HTTP/1.1",
+          "method": "GET",
+          "queryString": [],
+          "url": "https://fake_host/api/v1/namespaces/fake_ns/packages"
```
Comment (tripodsan) [inline test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4966]: how did you record this?

File: src/index.js [hunk 6]
```diff
@@ -27,6 +27,9 @@ async function publishConfig(params) {
     params.__ow_logger,
     params.indexconfig,
     params.algoliaappid,
+    params.wskAuth,
+    params.wskHost,
```
Comment (tripodsan) [inline src/index.js:31]: this should default to `process.env.__OW_API_HOST`

File: src/check-pkgs.js (hunk 2 above)
Comment (MarquiseRosier) [reply src/check-pkgs.js:26]: I wanted to have it resolved by the call to Promise.all

File: test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har (hunk 5 above)
Comment (MarquiseRosier) [reply test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4966]: I first recorded call to openwhisk with my credentials; and then put in all the fake stuff in the recordings

File: src/index.js (hunk 6 above)
Comment (MarquiseRosier) [reply src/index.js:31]: ^ Right! Good call, i'll fix this asap tomorrow morning

File: src/check-pkgs.js (hunk 2 above)
Comment (MarquiseRosier) [reply src/check-pkgs.js:26]: I don't think it'll lead to a race condition, I checked and check-pkgs always resolves first. so it'll cause the entire Promise.all to fail before anything is written to the VCL/dictionaries, if you think it's not safe for racing reasons, I can move it. and use the await notation as you advise to ensure it executes first.

Comment (tripodsan) [reply src/check-pkgs.js:26]: I don't understand...`Promise.all()` doesn't guarantee the order of when the promises are resolved. however, this is not the point.

the `new Promise()` here is returned.... this is the same as writing:

```js
async function check_pkgs(auth, host, namespace, config) {
...
```

File: test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har (hunk 5 above)
Comment (tripodsan) [reply test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4966]: right...but why change the host ?

Comment (MarquiseRosier) [reply test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4966]: right, that was probably overkill. Would it be any trouble to leave it? The authorization is fake

Comment (tripodsan) [reply test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4966]: sure...just keep it...I was just curious.

File: src/check-pkgs.js (hunk 2 above)
Comment (MarquiseRosier) [reply src/check-pkgs.js:26]: arghhh duhhh it always returns a promise! I will fix this.

File: src/check-pkgs.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+async function getPackageList(auth, host, namespace) {
+  let packageList;
+  const ow = openwhisk({
+    api_key: auth,
+    apihost: host,
+    namespace,
+  });
+
+  try {
+    packageList = await ow.packages.list();
```
Comment (tripodsan) [inline src/check-pkgs.js:50]:
```suggestion
    const packageList = await ow.packages.list();
```

File: src/check-pkgs.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    return packageList.reduce((prev, curr) => {
+      // eslint-disable-next-line no-param-reassign
+      prev[curr.name] = true;
+      return prev;
+    }, {});
+  } catch (e) {
+    throw new WhiskError('whisk failed to obtain package list');
```
Comment (tripodsan) [inline src/check-pkgs.js:57]: maybe log the error?

File: src/check-pkgs.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  }
+}
+
+/**
+ * @param auth {string} openwhisk authentication key
+ * @param host {string} openwhisk api host
+ * @param namespace {string} openwhisk namespace under which actions exist
+ * @param config {object} a Helix Configuration object
+ *
+ * This function checks that packages defined for
```
Comment (tripodsan) [inline src/check-pkgs.js:67]: convention is to have the description before the params.

File: src/check-pkgs.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  }
+}
+
+/**
+ *
+ * @param auth {string} openwhisk authentication key
+ * @param host {string} openwhisk api host
+ * @param namespace {string} openwhisk namespace under which actions exist
+ *
+ * This function gets list of packages deployed to OpenWhisk
```
Comment (tripodsan) [inline src/check-pkgs.js:39]: convention is to have the description before the params.

File: src/check-pkgs.js (hunk 7 above)
Comment (MarquiseRosier) [reply src/check-pkgs.js:50]: True, I left it because the error will get logged in the publish function. 

https://github.com/adobe/helix-publish/blob/331fb33b81761854e60950651189b25abd8e671c/src/publish.js#L98

File: src/check-pkgs.js (hunk 8 above)
Comment (MarquiseRosier) [reply src/check-pkgs.js:57]: True, I left it because the error will get logged in the publish function. 

https://github.com/adobe/helix-publish/blob/331fb33b81761854e60950651189b25abd8e671c/src/publish.js#L96-L105

File: src/index.js (hunk 6 above)
Comment (MarquiseRosier) [reply src/index.js:31]: I think this will happen in the publish command in helix-cli!

Comment (MarquiseRosier) [reply src/index.js:31]: https://github.com/adobe/helix-cli/blob/318770004ea81dd88bc825b73fc124343f24efd0/src/remotepublish.cmd.js#L285-L326

File: src/check-pkgs.js (hunk 8 above)
Comment (tripodsan) [reply src/check-pkgs.js:55]: I mean, log the reason for the error... the `e`

File: package.json (hunk 1 above)
Comment (trieloff) [reply package.json:93]: Just a guess: because it's pre-installed on the container but needed during testing. Making it a regular dependency would be cleaner, though.

File: src/check-pkgs.js (hunk 8 above)
Comment (tripodsan) [reply src/check-pkgs.js:55]: @MarquiseRosier so...will you add the error logging here?

File: package.json (hunk 1 above)
Comment (MarquiseRosier) [reply package.json:93]: I think this is outdated now

Comment (tripodsan) [reply package.json:93]: it's about `openwhisk`

File: test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har [hunk 11]
```diff
@@ -4930,6 +4930,105 @@
...
+              "name": "accept",
+              "value": "application/json"
+            },
+            {
+              "name": "user-agent",
+              "value": "openwhisk-client-js"
+            },
+            {
+              "name": "authorization",
+              "value": "Basic ZmFrZV9hdXRo"
```
Comment (trieloff) [inline test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4951]: This reads `fake_auth`. 😥

File: src/index.js [hunk 12]
```diff
@@ -27,6 +27,9 @@ async function publishConfig(params) {
...
+    params.wskNamespace,
```
Comment (tripodsan) [inline src/index.js:32]: please add fallback to:

- `process.env.__OW_API_KEY`
- `process.env.__OW_NAMESPACE`
- `process.env.__OW_API_HOST`

then, we don't need to pass the creds in `helix-cli`. see https://github.com/adobe/helix-cli/issues/1378#issuecomment-616878803

File: test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har (hunk 11 above)
Comment (MarquiseRosier) [reply test/fixtures/recordings/Integration-Test_519117731/Test-publish-function-locally_1268185746/recording.har:4951]: haha disappointed hacker

File: .releaserc.js [hunk 13]
```diff
@@ -3,7 +3,7 @@ module.exports = {
     '@semantic-release/commit-analyzer',
     '@semantic-release/release-notes-generator',
     ['@semantic-release/changelog', {
-      changelogFile: 'CHANGELOG.md',
+      chadangelogFile: 'CHANGELOG.md',
```
Comment (tripodsan) [inline .releaserc.js:6]: ????

**Comment by MarquiseRosier** [comment]
sweet thanks a ton @tripodsan!
